components/public-api/typescript-common/src/public-api-converter.ts

```diff
@@ -118,45 +129,122 @@ export class PublicAPIConverter {
     toWorkspace(arg: WorkspaceInfo | WorkspaceInstance, current?: Workspace): Workspace {
         const workspace = current ?? new Workspace();
 
+        workspace.spec = this.toWorkspaceSpec(arg, workspace.spec);
+        workspace.status = this.toWorkspaceStatus(arg, workspace.status);
+        workspace.metadata = this.toWorkspaceMetadata(arg, workspace.metadata);
         if ("workspace" in arg) {
             workspace.id = arg.workspace.id;
-            workspace.prebuild = arg.workspace.type === "prebuild";
-            workspace.organizationId = arg.workspace.organizationId;
-            workspace.name = arg.workspace.description;
-            workspace.pinned = arg.workspace.pinned ?? false;
-
-            const contextUrl = ContextURL.normalize(arg.workspace);
-            if (contextUrl) {
-                workspace.contextUrl = contextUrl;
+            if (arg.latestInstance) {
+                return this.toWorkspace(arg.latestInstance, workspace);
             }
+            return workspace;
+        }
+        return workspace;
+    }
+
+    toWorkspaceSpec(arg: WorkspaceInfo | WorkspaceInstance, current?: WorkspaceSpec): WorkspaceSpec {
+        const spec = new WorkspaceSpec(current);
+        if ("workspace" in arg) {
+            spec.admission = this.toAdmission(arg.workspace.shareable);
+            spec.initializer = new WorkspaceInitializer(spec.initializer);
+            spec.type =
+                arg.workspace.type === "prebuild"
+                    ? WorkspaceSpec_WorkspaceType.PREBUILD
+                    : WorkspaceSpec_WorkspaceType.REGULAR;
+            spec.environmentVariables = this.toEnvironmentVariables(arg.workspace.context);
+
             if (WithPrebuild.is(arg.workspace.context)) {
-                workspace.prebuildId = arg.workspace.context.prebuildWorkspaceId;
+                const initializerSpec = new WorkspaceInitializer_Spec({
+                    spec: {
+                        case: "prebuild",
+                        value: new PrebuildInitializer({
+                            prebuildId: arg.workspace.context.prebuildWorkspaceId,
+                        }),
+                    },
+                });
+                spec.initializer.specs.push(initializerSpec);
+            } else if (CommitContext.is(arg.workspace.context)) {
+                const initializerSpec = new WorkspaceInitializer_Spec({
+                    spec: {
+                        case: "git",
+                        value: new GitInitializer({
+                            remoteUri: arg.workspace.context.normalizedContextURL,
+                            upstreamRemoteUri: arg.workspace.context.upstreamRemoteURI,
+                            // TODO:
```

Comment:
are those TODOs relevant?

Reply:
No, we added lots of things in `message Workspace`, and it's the shape for the future (in next-gen). Add `TODOs` because we can't do convert currently

Reply:
We will create issue for them https://github.com/gitpod-io/gitpod/pull/19138#discussion_r1409033162